Add OAuth server

komret opened the pull request:
Implements OAuth flow for desktop apps as described [here](https://developers.google.com/identity/protocols/oauth2/native-app). We originally used this flow in desktop, but the app was registered as an IOS app as Google recommends in [docs](https://github.com/googleapis/google-auth-library-nodejs#oauth2-with-installed-apps-electron) of the library we are using. This allowed us to use the flow without the need to use `client_secret`. However, this approach is deprecated and there will be a user-facing warning from August as described [here](https://developers.googleblog.com/2022/02/making-oauth-flows-safer.html#disallowed-loopback). To keep the `client_secret` in secret, we need to proxy the authorization requests through a server which holds it.

Suggestions for a lighter solution are welcome.

- [x] Make this work locally with testing credentials
- [x] Fix `access_token` refresh
- [ ] Would be nice to prevent duplicate auth requests when connecting to the provider
- [ ] Set up a remote server hosting the app. There should be an `.env` file holding the client secret for Desktop App which can be obtained in [Google Cloud Platform](https://console.cloud.google.com/apis/credentials)
- [x] Remove testing credentials from the app

mroz22 commented on packages/auth-server/package.json:
```diff
@@ -0,0 +1,19 @@
+        "test:unit": "jest --passWithNoTests",
+        "type-check": "tsc --build",
+        "build:lib": "rimraf ./lib && tsc --build tsconfig.lib.json",
+        "dev": "node --loader ts-node/esm ./src/index.ts"
```
we are going to need "start" script implementing something like this https://www.npmjs.com/package/nodemon

komret replied:
956e6330f68c4858d6e63487f4954b53bcc3b90d

mroz22 commented on packages/auth-server/src/index.ts:
```diff
@@ -0,0 +1,65 @@
+            }),
+            method: 'POST',
+        });
+        const json = await response.json();
```
I would add validation of whether response is in expected format here and if not I would return some anticipated error

komret replied:
678e53853979e50eb07e697bec95930df99842d0

mroz22 commented on packages/auth-server/src/index.ts:
```diff
@@ -0,0 +1,65 @@
+const app = express();
+app.use(express.json());
+
+const port = 3005;
```
port for production should not be hardcoded.

komret replied:
678e53853979e50eb07e697bec95930df99842d0

tsusanka commented:
Just as a side note: @mroz22 please let's discuss this before deploying it anywhere.

mroz22 commented:
> Just as a side note: @mroz22 please let's discuss this before deploying it anywhere.

Sure. Basically our idea is to make it work even without it. But in case we will manage to deploy this backend one day, it will only improve user experience seamlessly

komret commented:
@mroz22 the changes I made today:
- rebased from develop to match the `node` versions
- removed `google-auth-library` because we customized it to an extent that it is not needed anymore
- split `Client.token` between `Client.refreshToken` and `Client.refreshToken` to prevent confusion which token is assigned there
- disabled saving `access_token` into IndexedDB because it is only temporary

When using implicit flow, there is now a difference from the current production behaviour:
1. Because the access token is no longer saved to storage, if a user authenticates and then returns to the app (or refreshes the page) within a hour, they will need to re-authenticate. I think this is a minor issue, especially once we implement the auth server. Saving the access token to the database seems weird, but it can be done easily if desired.
2. The implicit flow currently fails silently when a user returns to the app when their access token in storage is no longer valid. When they attempt to edit a label, they get an error until they reconnect to Google via settings. In this PR, user gets an error when the app loads and the token is then removed from storage so that they can authenticate again and edit the labels. I think this is an improvement, but the initial error toast is annoying, we could skip displaying the error...

komret commented:
> 1. Because the access token is no longer saved to storage, if a user authenticates and then returns to the app (or refreshes the page) within a hour, they will need to re-authenticate. I think this is a minor issue, especially once we implement the auth server. Saving the access token to the database seems weird, but it can be done easily if desired.

After discussing it with @mroz, I added `accessToken` to IndexedDB storage. I added a migration. I also rebased and squashed some commits.

matejkriz commented on packages/suite-desktop/src-electron/libs/http-receiver.ts:
```diff
@@ -230,7 +235,14 @@ export class HttpReceiver extends EventEmitter {
-        const template = this.applyTemplate('You may now close this window.');
+        const script = `
+            <script>
+                if (window.location.href.includes('#')) {
```
pretty please add a comment here why this is needed

komret replied:
Added a comment in 07c05e195eccebe757d88a0b84e4dbe314b968c8.

matejkriz commented on packages/suite-desktop/src-electron/libs/http-receiver.ts:
```diff
@@ -138,6 +141,7 @@ export class HttpReceiver extends EventEmitter {
      * Entry point for handling requests
      */
     private onRequest = (request: http.IncomingMessage, response: http.ServerResponse) => {
+        console.log('request.url', request.url);
```
forgotten?

komret replied:
Removed in 07c05e195eccebe757d88a0b84e4dbe314b968c8.

matejkriz commented:
`metadataActions.test.ts` needs to be fixed
https://github.com/trezor/trezor-suite/runs/7231989690?check_suite_focus=true#step:12:79

mroz22 commented on packages/suite/src/actions/suite/metadataActions.ts:
```diff
@@ -46,12 +47,12 @@ export type MetadataAction =
     switch (type) {
         case 'dropbox':
-            return new DropboxProvider(token);
+            return new DropboxProvider(tokens?.accessToken);
```
are we sure we don't need to pass refresh_token here in some cases?

komret replied:
There is only one token saved, which is the refresh token. Fixed in 07c05e195eccebe757d88a0b84e4dbe314b968c8.

mroz22 commented on packages/suite/src/storage/migrations/index.ts:
```diff
@@ -454,15 +455,35 @@ export const migrate: OnUpgradeFunc<SuiteDBSchema> = async (
                     discoveryStore.add(discovery);
                 });
             });
+
```
this needs version 29. version 28 is already in the wild thus it would not migrate db for people who already have it

komret replied:
Done in b19f1d28478f2a987585f43cc00f81341797f37c

mroz22 commented on packages/suite/src/storage/CHANGELOG.md:
```diff
@@ -3,6 +3,7 @@
 # 28
 
 -   with advent of connect v9, device.state field has changed. migrated all affected data.
+-   split to `token` property for storing OAuth tokens into `tokens.accessToken` and `tokens.refreshToken`
```
please move it to version 29

komret replied:
Done in b19f1d28478f2a987585f43cc00f81341797f37c

mroz22 commented on packages/suite/src/services/suite/metadata/GoogleProvider.ts:
```diff
@@ -126,7 +129,7 @@ class GoogleProvider extends AbstractMetadataProvider {
     handleProviderError(err: any) {
         // collect human readable errors from wherever possible or fill with own general message;
-        let message: string = err?.error?.message || err?.message;
+        let message = err?.error?.message || err?.message || err?.description || '';
```
what was the reason for this change :thinking:  ?

komret replied:
It popped up during rebase, I don't remember writing this, should I revert to the original?

mroz22 replied:
yes please :+1:

mroz22 commented on packages/suite/src/services/suite/metadata/DropboxProvider.ts:
```diff
@@ -175,7 +175,9 @@ class DropboxProvider extends AbstractMetadataProvider {
                 isCloud: this.isCloud,
-                token,
+                tokens: {
+                    refreshToken: token,
```
looks confusing, checkout my comment in packages/suite/src/actions/suite/metadataActions.ts 
accesss token passed from metadataActions here being assigned as refreshToken :thinking:

komret replied:
This should be a refresh token, fixed the error in metadataActions in 07c05e195eccebe757d88a0b84e4dbe314b968c8.
